scripts/enumerate_load_table_lookup.py

```diff
@@ -0,0 +1,377 @@
+# MODEL FUNCS
+
+
+def copydirectory(src, dst, override=False):
+    """
+    Copy directory from src to dst, with option to override.
+    override=True will clear dst before copying
+    """
+    if (override == False) & (os.path.exists(dst)):
+        logger.info(f'"{dst}" already exists, passing...')
+        pass
+
+    else:
+        if os.path.exists(dst):
+            shutil.rmtree(dst)
+        try:
+            shutil.copytree(src, dst)
+        except OSError as exc:
+            if exc.errno in (errno.ENOTDIR, errno.EINVAL):
+                shutil.copy(src, dst)
+            else:
+                raise
+        logger.info(f'"{src}"\n copied to: "{dst}"')
+
+
+@timed_info
+def relocate_original_file(lookup_file):
+    """ copy load_data_lookup to new location """
+    # define path for relocation
+    file_rename = list(os.path.splitext(os.path.basename(lookup_file)))
+    file_rename[0] = file_rename[0] + "_orig"
+    file_rename = "".join(file_rename)
+    relocated_file = os.path.join(os.path.dirname(lookup_file), file_rename)
+
+    # execute
+    copydirectory(lookup_file, relocated_file, override=False)
+    return relocated_file
+
+
+@timed_info
+def enumerate_lookup_by_keys(df_lookup, keys):
+    if len(keys) > 1:
+        df_lookup_full = df_lookup.select(keys[0]).distinct()
+
+        for key in keys[1:]:
+            df_lookup_full = df_lookup_full.crossJoin(df_lookup.select(key).distinct())
+
+        df_lookup_full = df_lookup_full.join(df_lookup, keys, "left").sort(["id"] + keys)
+    else:
+        df_lookup_full = df_lookup
+    return df_lookup_full
+
+
+@timed_info
+def enumeration_report(df_lookup_full, df_lookup):
+    N_df_lookup = df_lookup.count()
+    N_df_lookup_full = df_lookup_full.count()
+    N_df_lookup_null = N_df_lookup_full - N_df_lookup
+    logger.info(f"  # rows in df_lookup: {N_df_lookup}")
+    logger.info(f"  # rows in df_lookup (fully enumerated): {N_df_lookup_full}")
+    logger.info(
+        f"  # of rows without data: {N_df_lookup_null} ({(N_df_lookup_null/N_df_lookup_full*100):.02f}%)"
+    )
+
+
+@timed_info
+def assertion_checks(df_lookup_full, df_lookup, keys):
+
+    # 1) number of (data) id list is the same before and after enumeration
+    df_lookup_ids = df_lookup.select("id").distinct().toPandas().iloc[:, 0].values
+    df_lookup_full_ids = df_lookup_full.select("id").distinct().toPandas().iloc[:, 0].values
+
+    assert len(set(df_lookup_ids).difference(df_lookup_full_ids)) == 0
+
+    # 2) make sure N_df_lookup_full is the product of the nunique of each key
+    N_df_lookup_full = df_lookup_full.count()
+    N_enumerations = 1
+    for key in keys:
+        N_enumerations *= df_lookup.select(key).distinct().count()
+
+    assert N_enumerations == N_df_lookup_full
+
+
+def get_data_size(df, bytes_per_cell=64):
+    """ approximate dataset size """
+    n_rows = df.count()
+    n_cols = len(df.columns)
+    data_MB = n_rows * n_cols * bytes_per_cell / 1e6  # MiB
```

Comment:
For clarity,
Megabyte = MB = 1e6 = 1,000,000
Mibibyte = MiB = 2^20 = 1,048,576

Any software tool or device could report either variety. Storage manufacturers like MB because the number is bigger.